### PR TITLE
Add support for dataset,runtime,webhook pprof

### LINKED ISSUE #780: [FEATURES] Instant live visualization of Fluid  statistics (GC, MemStats, etc.).
**What feature you'd like to add:**

Add Performance analysis to Fluid: https://github.com/arl/statsviz

**Why is this feature needed:**

### PR DESCRIPTION
**Pull request by wangyanghack**
### Ⅰ. Describe what this PR does
add instant live visualization of statistics for performance analysis
### Ⅱ. Does this pull request fix one issue?

fixes #780 

### Ⅴ. Special notes for reviews
I cannot decide on nodeport number. please give me some advice. thanks.

### CODE REVIEW

**Comment by cheyang** [comment]
Thanks very much for your great contributions. If you'd like to add documentation, it will be great!

**Comment by wangyanghack** [comment]
> Thanks very much for your great contributions. If you'd like to add documentation, it will be great!
Thanks very much for your response. I will add documentation recently.
And Can you give me some advice to decide which number nodeport use or configurable? thanks

**Comment by wangyanghack** [comment]
> I think you can make nodeport configurable and the default port could be 50000.
> Could you rebase you code so that it can fix the CI. Thanks.
I have fixed nodeport and passed the CI.

File: charts/fluid/fluid/templates/controller/alluxioruntime_service.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{{ if .Values.runtime.alluxio.statistics.enabled -}}
+apiVersion: v1
+kind: Service
+metadata:
+  name: fluid-pod-alluxioruntime-controller
```
Comment (yangyuliufeng) [inline charts/fluid/fluid/templates/controller/alluxioruntime_service.yaml:5]: It's not suitable to be 'fluid-pod-alluxioruntime-controller', how about 'fluid-alluxioruntime-controller-ststic'

File: charts/fluid/fluid/values.yaml [hunk 2]
```diff
@@ -40,6 +46,9 @@ runtime:
       image: registry.cn-shanghai.aliyuncs.com/jindofs/jindo-fuse:3.6.0
     controller:
       image: fluidcloudnative/jindoruntime-controller:v0.7.0-109a578
+    statistics:
+      enabled: false
+      nodePort: 50002
```
Comment (yangyuliufeng) [inline charts/fluid/fluid/values.yaml:51]: I think it's better to put it in runtime.controllerStatistics, not three pieces.
我觉得统一写一个runtime.controllerStatistics比较好，不要分三块

**Review by yangyuliufeng — CHANGES_REQUESTED** [review]
I have tested this pr and found no problems.
I suggest changing the naming of some files such as the svc name.
Also， I'm worried that it's unsafe to expose program indicators through HTTP.

File: charts/fluid/fluid/templates/controller/alluxioruntime_service.yaml (hunk 1 above)
Comment (wangyanghack) [reply charts/fluid/fluid/templates/controller/alluxioruntime_service.yaml:5]: Yes,I will fix.

File: charts/fluid/fluid/values.yaml (hunk 2 above)
Comment (wangyanghack) [reply charts/fluid/fluid/values.yaml:51]: Yes, it's better. I will fix.

File: cmd/dataset/main.go [hunk 3]
```diff
@@ -75,6 +79,8 @@ func init() {
 	_ = datav1alpha1.AddToScheme(scheme)
 
 	startCmd.Flags().StringVarP(&metricsAddr, "metrics-addr", "", ":8080", "The address the metric endpoint binds to.")
+	startCmd.Flags().StringVarP(&statisticsAddr, "statistics-addr", "", ":6060", "The address the application runtime statistics endpoint binds to.")
+	startCmd.Flags().BoolVarP(&enablePerformanceAnalysis, "enable-performance-analysis", "", false, "Enable performance analysis by instant live visualization of application runtime statistics")
```
Comment (wangyanghack) [inline cmd/dataset/main.go:83]: @yangyuliufeng Thanks very much for your review. I will fix all. 
By the way, which is better between enable-performance-analysis and enable-statistics-visualization here?

**Comment by wangyanghack** [comment]
> Also， I'm worried that it's unsafe to expose program indicators through HTTP.
@yangyuliufeng  @cheyang  I cannot decide where tls crt and key should be. 
My plan is :
1. execute a certificate.sh to generate tls crt and key
2. execute kubectl create secret tls tls-secret --cert=path/to/tls.cert --key=path/to/tls.key to generate a secret
3. pod of fluid volumemount this secret to use tls crt and key
But my plan has a problem: it cannot execute certificate.sh and create secret automatically when helm install.
Please give me some advice. Thanks very much.

File: charts/fluid/fluid/templates/controller/dataset_service.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{{ if .Values.dataset.controller.statistics.enabled -}}
```
Comment (cheyang) [inline charts/fluid/fluid/templates/controller/dataset_service.yaml:1]: Maybe the service is not necessary. I think we can connect to ip address of pod directly.

File: cmd/alluxio/main.go [hunk 5]
```diff
@@ -17,12 +17,15 @@ package main
 
 import (
 	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/arl/statsviz"
```
Comment (cheyang) [inline cmd/alluxio/main.go:23]: How about using "net/http/pprof" ? Kubernetes is also using this.

**Review by cheyang — CHANGES_REQUESTED** [review]

**Comment by cheyang** [comment]
> > Also， I'm worried that it's unsafe to expose program indicators through HTTP.
> 
> @yangyuliufeng @cheyang I cannot decide where tls crt and key should be.
> My plan is :
> 
> 1. execute a certificate.sh to generate tls crt and key
> 2. execute kubectl create secret tls tls-secret --cert=path/to/tls.cert --key=path/to/tls.key to generate a secret
> 3. pod of fluid volumemount this secret to use tls crt and key
> 
> But my plan has a problem: it cannot execute certificate.sh and create secret automatically when helm install.
> Please give me some advice. Thanks very much.

Thank you for your contribution!  My suggestion is that if we can leverage the implementation in Kubernetes. https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/server/routes/profiling.go#L30

**Comment by cheyang** [comment]
> I cannot decide on nodeport number. please give me some advice. thanks.

I think service to expose is not necessary for debugging.

File: cmd/alluxio/main.go (hunk 5 above)
Comment (wangyanghack) [reply cmd/alluxio/main.go:23]: Thanks very much for your good advice. I will use pprof to implement it.

**Comment by cheyang** [comment]
@wangyanghack  could you refer to https://github.com/banzaicloud/logging-operator/blob/master/main.go#L122-L129? thanks in advance!

**Comment by wangyanghack** [comment]
> @wangyanghack could you refer to https://github.com/banzaicloud/logging-operator/blob/master/main.go#L122-L129? thanks in advance!

Thank you very much for your careful guidance and patience. 
I have followed your advice to add pprof. Please review it and thanks a lot.

**Comment by wangyanghack** [comment]
@cheyang please task a moment to review my code. I will close it if it is not still satisfied with you again. Thanks a lot.

**Comment by cheyang** [comment]
@wangyanghack Please fix the conflict, thanks.

**Comment by wangyanghack** [comment]
@cheyang hi, please review it when you are free. Thanks very much.

File: cmd/alluxio/main.go [hunk 6]
```diff
@@ -116,6 +123,41 @@ func handle() {
 		}
 	}))
 
+	if pprofAddr != "" {
+		glog.Infof("Enabling pprof with address %s", pprofAddr)
```
Comment (cheyang) [inline cmd/alluxio/main.go:127]: Can we use the same log system in the controller. I suggest not to use `glog` here rahter than `github.com/go-logr/logr`.

**Review by cheyang — CHANGES_REQUESTED** [review]

File: cmd/dataset/main.go [hunk 7]
```diff
@@ -107,6 +114,41 @@ func handle() {
 		}
 	}))
 
+	if pprofAddr != "" {
```
Comment (cheyang) [inline cmd/dataset/main.go:117]: How about to create a common function to launch pprof?  This piece of code has been repeated several times. Just a suggestion, WDYT?

Comment (wangyanghack) [reply cmd/dataset/main.go:117]: yes, I will create a common function now

File: cmd/alluxio/main.go (hunk 6 above)
Comment (wangyanghack) [reply cmd/alluxio/main.go:127]: ok

File: pkg/utils/pprof.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
+package utils
```
Comment (cheyang) [inline pkg/utils/pprof.go:1]: Please add license header. Thanks.

File: pkg/utils/pprof.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+		mux.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
+		mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
+		mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+		mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
+
+		pprofServer := http.Server{
+			Addr:    pprofAddr,
+			Handler: mux,
+		}
+		glog.Infof("Starting pprof HTTP server at %s", pprofServer.Addr)
```
Comment (cheyang) [inline pkg/utils/pprof.go:26]: Please use logr instead of glog. Thanks.

**Review by xieydd — CHANGES_REQUESTED** [review]
@wangyanghack We can add test for new func to improve the diff hit of codecov, thanks.

**Review by cheyang — APPROVED** [review]

**Pull request merged**